integration/spadev0/main.go

```diff
@@ -0,0 +1,200 @@
+
+			// logging.SetLogLevel("spade-v0-tasks", "DEBUG")
+
+			// Extract the sources from the flag
+			sources := cctx.StringSlice("sources")
+
+			for _, source := range sources {
+				res, err := fetchActiveReplicas(ctx, source)
+
+				if err != nil {
+					return err
+				}
+
+				var perProvider = make(map[int]ProviderReplicas)
+
+				for _, replica := range res.ActiveReplicas {
+					for _, contract := range replica.Contracts {
+						providerID := contract.ProviderID
+						size := 2 << replica.PieceLog2Size % 30 // Convert to GiB
+						perProvider[providerID] = ProviderReplicas{
+							size: perProvider[providerID].size + size,
+							replicas: append(perProvider[providerID].replicas, Replica{
+								PieceCID:        replica.PieceCID,
+								PieceLog2Size:   replica.PieceLog2Size,
+								OptionalDagRoot: replica.OptionalDagRoot,
+							}),
+						}
+					}
+				}
+
+				replicasToTest := selectReplicasToTest(perProvider)
+
+				totalCids := 0
+				totalSize := 0
+				for prov, rps := range replicasToTest {
+					provider := perProvider[prov]
+					logger.Debugf("provider %d is storing %d GiB will have %d tests\n", prov, provider.size, len(rps))
+					totalCids += len(rps)
+					totalSize += provider.size
+				}
+
+				logger.Debugf("total %d CIDs will be tested for %d providers\n", totalCids, len(replicasToTest))
+
+				AddSpadeTasks(ctx, "spadev0", replicasToTest)
+			}
+			return nil
+		},
+	}
+
+	if err := app.Run(os.Args); err != nil {
+		panic(err)
+	}
+}
+
+func fetchActiveReplicas(ctx context.Context, url string) (*ActiveReplicas, error) {
+	logger.Debugf("fetching CIDs from %s", url)
+
+	req, err := http.NewRequestWithContext(ctx,
+		http.MethodGet,
+		url,
+		nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create request")
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to make request")
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("failed to get spade CID list: %s", resp.Status)
+	}
+
+	decompressor, err := zstd.NewReader(resp.Body)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create decompressor")
+	}
+
+	defer decompressor.Close()
+
+	data, err := ioutil.ReadAll(decompressor)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read decompressed data")
+	}
+
+	// Unmarshal the JSON data into your struct
+	var activeReplicas ActiveReplicas
+	err = json.Unmarshal(data, &activeReplicas)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal JSON data")
+	}
+
+	return &activeReplicas, nil
+}
+
+// Compute a number of CIDs to test, based on the total size of data (assuming in GiB)
+// Minimum 1, then log2 of the size in TiB
+// ex:
+// < 4TiB = 1 cid
+// 4 TiB - 16TiB = 2 cids
+// 16 TiB - 32 TiB = 3 cids
+// 32 TiB - 64 TiB = 4 cids
+// 64 TiB - 128 TiB = 5 cids
+// 128 TiB - 256 TiB = 6 cids
+// etc...
+func numCidsToTest(size int) int {
```

Comment:
@ribasushi here is my naive approach to making sure that all providers have at least one retrieval test, while scaling up the CIDs we test as they store more bytes. Let me know if this is fine for v0.

I did it based on GiB stored, not # of replicas stored, in order to avoid biasing the testing towards SPs who are storing lots of small CIDs vs. few big ones (I know they're all the same size 32GiB as of now, but just to be comprehensive) 

These are just the "top level" cids of the tree. Once the task gets picked up by the retrieval bot, we'll have an additional selection from the bottom layer of the tree for the actual data CIDs.

Here is the expected output running it against the current `active replicas`:

```bash
provider 440429 is storing 41040 GiB will have 5 tests
provider 33356 is storing 12224 GiB will have 3 tests
provider 1137229 is storing 27712 GiB will have 4 tests
provider 30379 is storing 53456 GiB will have 5 tests
provider 20378 is storing 75744 GiB will have 6 tests
provider 1163272 is storing 4336 GiB will have 2 tests
provider 1392893 is storing 394752 GiB will have 8 tests
provider 1806491 is storing 34544 GiB will have 5 tests
provider 10202 is storing 185440 GiB will have 7 tests
provider 717969 is storing 354912 GiB will have 8 tests
provider 1847751 is storing 46176 GiB will have 5 tests
provider 2576 is storing 656 GiB will have 1 tests
provider 1889910 is storing 56848 GiB will have 5 tests
provider 104671 is storing 4944 GiB will have 2 tests
provider 1771403 is storing 331920 GiB will have 8 tests
provider 1746964 is storing 11920 GiB will have 3 tests
provider 8399 is storing 17728 GiB will have 4 tests
provider 1784458 is storing 244976 GiB will have 7 tests
provider 1611097 is storing 128 GiB will have 1 tests
provider 97777 is storing 64416 GiB will have 5 tests
provider 3488 is storing 7216 GiB will have 2 tests
provider 1278 is storing 33216 GiB will have 5 tests
provider 2095132 is storing 375280 GiB will have 8 tests
provider 8403 is storing 13296 GiB will have 3 tests
provider 1310564 is storing 15504 GiB will have 3 tests
provider 2033496 is storing 194320 GiB will have 7 tests
provider 773157 is storing 8000 GiB will have 2 tests
total 124 cids will be tested for 27 providers
```

Reply:
The idea being we'd run this Integration once a day to make sure everyone is getting tested